### ISSUE TITLE
Resolve confusion between `classification.Dice` and `segmentation.DiceScore` and improve the implementation

### ISSUE THREAD

**Issue by nkaenzig** [Lightning-AI/torchmetrics#3014]
## 🚀 Feature

Resolve confusion around dice score implementations and improve the current implementation.

### Motivation

Currently there are two different implementations for dice in `torchmetrics`:
- `torchmetrics.classification.Dice`
- `torchmetrics.segmentation.DiceScore`

It seems that originally the idea was to deprecate the old `Dice` implementation, as mentioned in https://github.com/Lightning-AI/torchmetrics/pull/2725.

Not only is the coexistence of those two implementations confusing, there are also significant differences in the implementation of the two metrics, which can lead to significantly different scores in downstream tasks, especially in multi-class settings. For instance in `macro` mode, `DiceScore` by default calculates the pixel overlaps/unions first separately per sample and then when doing the division `numerator/denominator` uses a fallback value of `zero_division` (https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/functional/segmentation/dice.py#L94). `Dice` on the other hand doesn't calculate per-sample (see `mdmc_average="global"`) and excludes entries where the denominator is 0 (https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/functional/classification/dice.py#L49).

Most of those differences stem from different possible strategies of dealing with 0 values in the denominator of the dice metric. A few possible choices here are:
1. Ignore samples/entries where denominator is 0 (this is what `classification.Dice` does in `macro` mode)
2. Ignore samples where gnd truth is empty (this is what `monai.metrics.DiceMetric` does by default, which can be controlled through the `ignore_empty` init argument)
3. Ignore samples where both the prediction and the ground truth is empty (i.e. the denominator for that sample is all 0).
4. Use `1.0` or `0.0` fallback values for `0` divisions (`segmentation.DiceScore` uses a `1.0` default fallback for `torchmetrics` versions `<=1.6.1` and a `0.0` fallback value for version `1.6.2`)
5. Always use `nan` as fallback value combined with `torch.nanmean` and `torch.nansum` aggregation functions
6. Before doing the `numerator/denominator` divison, do a `torch.sum(numerator, dim=0)`, `torch.sum(denominator, dim=0)`, summing up the pixel overlap & union counts accross all samples. You then shouldn't have any 0 values in the denominator anymore, assuming that each class occurs at least once in one sample in the ground truth.

While each choice has its pros/cons, I do think that option 4. which `segmentation.DiceScore` currently implements is particularly problematic. Especially for rare classes, this pulls the aggregated score towards the fallback value, and as a consequence, the resulting score cannot be interpreted visually as "overlap" anymore, as it leads to either **overestimating** or **underestimating** the real overlap / score. To visualize this, here a real example, where a model yields a much higher aggregated macro dice score due to the fallback value being set to 1.0 (this experiment used `torchmetrics==1.6.1`):
<img width="1190" alt="Image" src="https://github.com/user-attachments/assets/d325c144-e069-4c63-9a52-494439fd14c5" />
Note that the orange model performs significantly worse than blue when looking at the classwise dice metrics. However, the aggregated metric is much higher, because the score here is driven towards the fallback value, which is `1.0` for `torchmetrics==1.6.1`.

### Pitch

- Deprecate `torchmetrics.classification.Dice` to avoid confusion between the two metrics, which also calculate different things.
- Improve the implementation of `torchmetrics.segmentation.Dice`. I recommend moving away from `0.0` / `1.0` fallback values. To me, option 3. makes the most sense (note that option 2. which MONAI implements and which is similar to 3. has the issue of not measuring false positives on samples where a class is not present at all). Also, implementing 6. can be added as an additional option by introducing a `aggregation_level: Literal["sample", "global"]` init argument. This can be easily implemented by adding `torch.sum` at the beginning of `_dice_score_compute`:
```python
    if aggregation_level == "global":
        numerator = torch.sum(numerator, dim=0)
        denominator = torch.sum(denominator, dim=0)
        support = torch.sum(support, dim=0) if support is not None else None
```

Given that dice is the main metric used for semantic segmentation tasks nowadays, I think improving this implementation would have a big impact both on research & industry usecases.

**Comment by SkafteNicki** [comment]
Hi @nkaenzig, thanks for raising this issue.
I created PR #3017 to remove Dice from classification. It has been deprecated since v1.6 and was set to be removed in v1.7 (next release). For the other parts of this issue, I take care of it later but I overall agree with your observations and that it should be corrected.

**Comment by nkaenzig** [comment]
@SkafteNicki Thanks for the feedback & quick action. I created a draft PR here: https://github.com/Lightning-AI/torchmetrics/pull/3018
Please let me know if you think this goes into the right direction.
